src/formatting/visitor.rs

```diff
@@ -181,18 +181,56 @@ impl<'b, 'a: 'b> FmtVisitor<'a> {
     }
 
     /// Remove spaces between the opening brace and the first statement or the inner attribute
-    /// of the block.
-    fn trim_spaces_after_opening_brace(
+    /// of the block, fast-forwarding this position of the visitor.
+    /// If `preserve_block_start_blank_lines` is true, the return value contains the newlines that
+    /// should be preserved.
+    pub(crate) fn trim_spaces_after_opening_brace(
         &mut self,
-        b: &ast::Block,
-        inner_attrs: Option<&[ast::Attribute]>,
-    ) {
-        if let Some(first_stmt) = b.stmts.first() {
-            let hi = inner_attrs
-                .and_then(|attrs| inner_attributes(attrs).first().map(|attr| attr.span.lo()))
-                .unwrap_or_else(|| first_stmt.span().lo());
-            let missing_span = self.next_span(hi);
+        first_item_pos: Option<BytePos>,
+    ) -> String {
+        let mut result = String::new();
+        if let Some(first_item_pos) = first_item_pos {
+            let missing_span = self.next_span(first_item_pos);
             let snippet = self.snippet(missing_span);
+
+            if self.config.preserve_block_start_blank_lines() {
+                // First we need to find the span to look for blank lines in. This is either the
+                // - span between the opening brace and first item, or
+                // - span between the opening brace and a comment before the first item
+                // We do this so that we get a span of contiguous whitespace, which makes
+                // processing the blank lines easier.
+                let blank_lines_snippet = if let Some(hi) = self
+                    .snippet_provider
+                    .span_to_snippet(missing_span)
+                    .and_then(|s| s.find('/'))
+                {
+                    self.snippet_provider.span_to_snippet(mk_sp(
+                        missing_span.lo(),
+                        missing_span.lo() + BytePos::from_usize(hi),
+                    ))
+                } else {
+                    self.snippet_provider.span_to_snippet(missing_span)
+                };
+
+                if let Some(snippet) = blank_lines_snippet {
+                    let has_multiple_blank_lines =
+                        if let (Some(l), Some(r)) = (snippet.find('\n'), snippet.rfind('\n')) {
+                            l != r
+                        } else {
+                            false
+                        };
+                    if has_multiple_blank_lines {
+                        let mut lines = snippet.lines().map(&str::trim);
+                        lines.next(); // Eat block-opening newline
+                        while let Some("") = lines.next() {
+                            result.push('\n');
+                        }
+                    }
+                } else {
+                    debug!("Failed to preserve blank lines for {:?}", missing_span);
```

Comment:
Did you encounter this scenario at all during your testing, or is this a defensive/just-in-case?

Reply:
No, just adding as a foresight